# Library Importation

In [59]:
import pandas as pd 
import numpy as np 
import plotly.express as px 
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
import statsmodels.api as sm

# Data Importation and Exploration

In [4]:
## import data into pandas dataframe 
df = pd.read_csv('/kaggle/input/student-stress-factors/Student Stress Factors.csv')

##checking the columns of the data 
columns = df.columns
columns

                

Index(['Timestamp', 'Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'How would you rate you academic performance 👩‍🎓?',
       'how would you rate your study load?',
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?'],
      dtype='object')

In [5]:
## checking the structure and data types of the datframe 
data_structure = df.info()
data_structure

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53 entries, 0 to 52
Data columns (total 7 columns):
 #   Column                                                            Non-Null Count  Dtype 
---  ------                                                            --------------  ----- 
 0   Timestamp                                                         53 non-null     object
 1   Kindly Rate your Sleep Quality 😴                                  53 non-null     int64 
 2   How many times a week do you suffer headaches 🤕?                  53 non-null     int64 
 3   How would you rate you academic performance 👩‍🎓?                  53 non-null     int64 
 4   how would you rate your study load?                               53 non-null     int64 
 5   How many times a week you practice extracurricular activities 🎾?  53 non-null     int64 
 6   How would you rate your stress levels?                            53 non-null     int64 
dtypes: int64(6), object(1)
memory usage: 3.0+ KB


In [6]:
## checking the summary of the dataframe
summary = df.describe()
summary

Kindly Rate your Sleep Quality 😴  \
count                         53.000000   
mean                           3.150943   
std                            1.199298   
min                            1.000000   
25%                            2.000000   
50%                            3.000000   
75%                            4.000000   
max                            5.000000   

       How many times a week do you suffer headaches 🤕?  \
count                                         53.000000   
mean                                           1.981132   
std                                            1.263246   
min                                            1.000000   
25%                                            1.000000   
50%                                            1.000000   
75%                                            3.000000   
max                                            5.000000   

       How would you rate you academic performance 👩‍🎓?  \
count                                         53.000000   
mean                                           3.226415   
std                                            1.154281   
min                                            1.000000   
25%                                            3.000000   
50%                                            3.000000   
75%                                            4.000000   
max                                            5.000000   

       how would you rate your study load?  \
count                            53.000000   
mean                              2.811321   
std                               1.428509   
min                               1.000000   
25%                               2.000000   
50%                               2.000000   
75%                               4.000000   
max                               5.000000   

       How many times a week you practice extracurricular activities 🎾?  \
count                                          53.000000                  
mean                                            2.886792                  
std                                             1.449939                  
min                                             1.000000                  
25%                                             2.000000                  
50%                                             3.000000                  
75%                                             4.000000                  
max                                             5.000000                  

       How would you rate your stress levels?  
count                               53.000000  
mean                                 2.792453  
std                                  1.377826  
min                                  1.000000  
25%                                  2.000000  
50%                                  3.000000  
75%                                  4.000000  
max                                  5.000000

In [7]:
## checking the first five(5) rows of the dataframe
first_5_rows = df.head(5)
first_5_rows

Timestamp  Kindly Rate your Sleep Quality 😴  \
0  27/10/2023 21:54:15                                 3   
1  28/10/2023 12:24:40                                 4   
2  28/10/2023 12:24:51                                 2   
3  28/10/2023 12:26:11                                 3   
4  28/10/2023 12:26:45                                 2   

   How many times a week do you suffer headaches 🤕?  \
0                                                 1   
1                                                 1   
2                                                 1   
3                                                 2   
4                                                 3   

   How would you rate you academic performance 👩‍🎓?  \
0                                                 3   
1                                                 2   
2                                                 2   
3                                                 3   
4                                                 1   

   how would you rate your study load?  \
0                                    4   
1                                    3   
2                                    1   
3                                    2   
4                                    5   

   How many times a week you practice extracurricular activities 🎾?  \
0                                                  2                  
1                                                  3                  
2                                                  4                  
3                                                  3                  
4                                                  5                  

   How would you rate your stress levels?  
0                                       3  
1                                       2  
2                                       4  
3                                       3  
4                                       3

# Data Cleaning and Validation

In [8]:
## checking for null values 
null_values = df.isnull().sum()
null_values

Timestamp                                                           0
Kindly Rate your Sleep Quality 😴                                    0
How many times a week do you suffer headaches 🤕?                    0
How would you rate you academic performance 👩‍🎓?                    0
how would you rate your study load?                                 0
How many times a week you practice extracurricular activities 🎾?    0
How would you rate your stress levels?                              0
dtype: int64

In [9]:
## checking for duplicated rows 
duplicated_rows = df.duplicated().sum()
duplicated_rows

0

# Data Exploration Analysis 

In [97]:
df

Timestamp  Kindly Rate your Sleep Quality 😴  \
0   27/10/2023 21:54:15                                 3   
1   28/10/2023 12:24:40                                 4   
2   28/10/2023 12:24:51                                 2   
3   28/10/2023 12:26:11                                 3   
4   28/10/2023 12:26:45                                 2   
5   28/10/2023 12:31:02                                 3   
6   28/10/2023 12:34:45                                 3   
7   28/10/2023 12:35:43                                 4   
8   28/10/2023 12:36:07                                 2   
9   28/10/2023 12:36:20                                 1   
10  28/10/2023 12:37:22                                 2   
11  28/10/2023 12:38:40                                 3   
12  28/10/2023 12:39:43                                 2   
13  28/10/2023 12:40:50                                 4   
14  28/10/2023 12:41:12                                 4   
15  28/10/2023 12:41:56                                 2   
16  28/10/2023 12:48:22                                 3   
17  28/10/2023 13:26:41                                 1   
18  28/10/2023 13:28:47                                 2   
19  28/10/2023 13:29:54                                 2   
20  28/10/2023 13:30:57                                 3   
21  28/10/2023 13:32:41                                 3   
22  28/10/2023 13:33:15                                 2   
23  28/10/2023 13:33:59                                 3   
24  28/10/2023 13:41:28                                 5   
25  28/10/2023 13:42:52                                 3   
26  28/10/2023 13:44:07                                 5   
27  28/10/2023 13:47:44                                 5   
28  28/10/2023 13:52:22                                 2   
29  28/10/2023 13:54:48                                 3   
30  28/10/2023 13:57:20                                 5   
31  28/10/2023 14:09:05                                 5   
32  28/10/2023 14:27:19                                 5   
33  28/10/2023 14:40:55                                 2   
34  28/10/2023 14:58:39                                 4   
35  28/10/2023 15:10:02                                 5   
36  28/10/2023 15:10:43                                 4   
37  28/10/2023 15:39:59                                 3   
38  28/10/2023 16:16:13                                 3   
39  28/10/2023 16:34:16                                 3   
40  28/10/2023 16:55:49                                 4   
41  28/10/2023 17:24:10                                 4   
42  28/10/2023 17:26:18                                 4   
43  28/10/2023 17:29:55                                 2   
44  28/10/2023 17:31:52                                 5   
45  28/10/2023 17:56:58                                 4   
46  28/10/2023 22:33:48                                 2   
47  28/10/2023 23:06:00                                 4   
48  28/10/2023 23:45:09                                 3   
49  29/10/2023 00:32:26                                 1   
50  29/10/2023 11:40:20                                 5   
51  31/10/2023 10:10:03                                 1   
52  31/10/2023 20:49:11                                 3   

    How many times a week do you suffer headaches 🤕?  \
0                                                  1   
1                                                  1   
2                                                  1   
3                                                  2   
4                                                  3   
5                                                  1   
6                                                  5   
7                                                  3   
8                                                  1   
9                                                  2   
10                                                 3   
11                                             

## Histogram for Sleep Quality Rating"

In [10]:
fig = px.histogram(df, x='Kindly Rate your Sleep Quality 😴')
fig.update_layout(
    title="Sleep Quality Rating",
    xaxis_title="Sleep Quality",
    yaxis_title="Frequency"
)

fig.show()

## Histogram for Days with Headaches 

In [11]:
fig = px.histogram(df, x='How many times a week do you suffer headaches 🤕?')
fig.update_layout(
    title="Days with Headaches",
    xaxis_title="Days with Headaches",
    yaxis_title="Frequency"
)

fig.show()

## Histogram for Academic Rating

In [12]:
fig = px.histogram(df, x='How would you rate you academic performance 👩‍🎓?')
fig.update_layout(
    title="Academic Perfomance Rating",
    xaxis_title="Academic Rating",
    yaxis_title="Frequency"
)

fig.show()


## Histogram for Study Load Ratings 

In [13]:
fig = px.histogram(df, x='how would you rate your study load?')
fig.update_layout(
    title="Study Load Rating",
    xaxis_title="Study Load Rating",
    yaxis_title="Frequency"
)

fig.show()


## Histogram for Extracurrilcular Activity

In [14]:
fig = px.histogram(df, x='How many times a week you practice extracurricular activities 🎾?')
fig.update_layout(
    title="Extracurrilcular Activity Rating",
    xaxis_title="Extracurrilcular Activity Rating",
    yaxis_title="Frequency"
)

fig.show()

## Histogram for Stress Level Rating 

In [15]:
fig = px.histogram(df, x='How would you rate your stress levels?')
fig.update_layout(
    title="Stress Level Rating",
    xaxis_title="Stress Level Rating",
    yaxis_title="Frequency"
)

fig.show()

# Correlation Analysis 

In [16]:
df.columns

Index(['Timestamp', 'Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'How would you rate you academic performance 👩‍🎓?',
       'how would you rate your study load?',
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?'],
      dtype='object')

In [17]:
correlation = df[['Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'how would you rate your study load?',
        'How would you rate you academic performance 👩‍🎓?',          
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?']].corr()
correlation

Kindly Rate your Sleep Quality 😴  \
Kindly Rate your Sleep Quality 😴                                            1.000000   
How many times a week do you suffer headaches 🤕?                            0.027303   
how would you rate your study load?                                         0.095518   
How would you rate you academic performance 👩‍🎓?                            0.266565   
How many times a week you practice extracurricu...                         -0.001043   
How would you rate your stress levels?                                      0.286995   

                                                    How many times a week do you suffer headaches 🤕?  \
Kindly Rate your Sleep Quality 😴                                                            0.027303   
How many times a week do you suffer headaches 🤕?                                            1.000000   
how would you rate your study load?                                                         0.104557   
How would you rate you academic performance 👩‍🎓?                                           -0.115711   
How many times a week you practice extracurricu...                                         -0.179676   
How would you rate your stress levels?                                                     -0.035439   

                                                    how would you rate your study load?  \
Kindly Rate your Sleep Quality 😴                                               0.095518   
How many times a week do you suffer headaches 🤕?                               0.104557   
how would you rate your study load?                                            1.000000   
How would you rate you academic performance 👩‍🎓?                               0.073057   
How many times a week you practice extracurricu...                             0.054481   
How would you rate your stress levels?                                         0.341232   

                                                    How would you rate you academic performance 👩‍🎓?  \
Kindly Rate your Sleep Quality 😴                                                            0.266565   
How many times a week do you suffer headaches 🤕?                                           -0.115711   
how would you rate your study load?                                                         0.073057   
How would you rate you academic performance 👩‍🎓?                                            1.000000   
How many times a week you practice extracurricu...                                          0.050081   
How would you rate your stress levels?                                                      0.005932   

                                                    How many times a week you practice extracurricular activities 🎾?  \
Kindly Rate your Sleep Quality 😴                                                            -0.001043                  
How many times a week do you suffer headaches 🤕?                                            -0.179676                  
how would you rate your study load?                                                          0.054481                  
How would you rate you academic performance 👩‍🎓?                                             0.050081                  
How many times a week you practice extracurricu...                                           1.000000                  
How would you rate your stress levels?                                                       0.180536                  

                                                    How would you rate your stress levels?  
Kindly Rate your Sleep Quality 😴                                                  0.286995  
How many times a week do you suffer headaches 🤕?                                 -0.035439  
how would you rate your study load?                                               0.341232  
How would you rate you academic performance 👩‍🎓?                                  0.005932  
How many times a week you practic

# Linear Regression Analysis 


## Splitting the data into X and Y 

In [18]:
df.columns

Index(['Timestamp', 'Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'How would you rate you academic performance 👩‍🎓?',
       'how would you rate your study load?',
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?'],
      dtype='object')

In [30]:
x = df[['Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'how would you rate your study load?',
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?']].to_numpy()

In [29]:
y = df['How would you rate you academic performance 👩‍🎓?'].to_numpy().reshape(-1,1)

## Training the model 

In [31]:
x_train,x_test,y_train,y_test = train_test_split(x,y, test_size = 0.2, random_state = 10)

In [75]:
model= LinearRegression()

## Fitting the Varables into the Model 

In [76]:
model.fit(x_train,y_train)

LinearRegression()

## Linear Regression Coefficient 

In [77]:
linear_coefficient = model.coef_
linear_coefficient

array([[ 0.19577561, -0.11549465,  0.02763273, -0.0420948 , -0.10584436]])

## Academic Performance Prediction 

In [78]:
performance_prediction = model.predict(x).astype(int)


## Coefficient Determination (r2)

In [79]:
coefficient_determination = r2_score(y,performance_prediction)
coefficient_determination

-0.19798474945533795

R-squared value of -0.19798474945533795 means that the model's predictions are not explaining much of the variance in academic performance and, in fact, may be performing poorly.

# Ordinary Least Square Model 

In [80]:
# Add a constant (intercept) to the independent variables
X = df[['Kindly Rate your Sleep Quality 😴',
       'How many times a week do you suffer headaches 🤕?',
       'how would you rate your study load?',
       'How many times a week you practice extracurricular activities 🎾?',
       'How would you rate your stress levels?']]

X = sm.add_constant(X)


In [81]:
# Create an OLS model
model = sm.OLS(y, X)

In [82]:
# Fit the model to your data
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.103
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1.078
Date:                Sun, 05 Nov 2023   Prob (F-statistic):              0.385
Time:                        06:06:18   Log-Likelihood:                -79.425
No. Observations:                  53   AIC:                             170.9
Df Residuals:                      47   BIC:                             182.7
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
====================================================================================================================================
                                                                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------------------------------------
const                                                                2.5306      0.671      3.770      0.000       1.180       3.881
Kindly Rate your Sleep Quality 😴                                     0.2856      0.139      2.054      0.046       0.006       0.565
How many times a week do you suffer headaches 🤕?                    -0.1194      0.129     -0.923      0.361      -0.380       0.141
how would you rate your study load?                                  0.0798      0.120      0.667      0.508      -0.161       0.321
How many times a week you practice extracurricular activities 🎾?     0.0352      0.114      0.309      0.758      -0.194       0.264
How would you rate your stress levels?                              -0.1052      0.130     -0.807      0.423      -0.367       0.157
==============================================================================
Omnibus:                        0.873   Durbin-Watson:                   1.692
Prob(Omnibus):                  0.646   Jarque-Bera (JB):                0.828
Skew:                          -0.287   Prob(JB):                        0.661
Kurtosis:                       2.788   Cond. No.                         27.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

 ## Key points to note include:

R-squared (R^2): This is a measure of how well the model explains the variance in the dependent variable. In this case, the R-squared is 0.103, indicating that the independent variables explain approximately 10.3% of the variance in the dependent variable.


F-statistic and Prob (F-statistic): The F-statistic tests the overall significance of the model. In this case, the F-statistic is 1.078 with a probability of 0.385, which suggests that the model may not be statistically significant.


Coefficients (coef): These values represent the estimated coefficients for each independent variable in the model. The standard errors, t-statistics, and p-values are also provided to assess the statistical significance of each coefficient.


AIC and BIC: These are information criteria used for model selection. Lower values indicate a better-fitting model.


Omnibus, Durbin-Watson, Jarque-Bera, Skew, and Kurtosis: These are various statistical tests and measures related to the model's assumptions and fit. For example, the Durbin-Watson value checks for autocorrelation in the residuals.


Cond. No.: This represents the condition number, which can indicate potential multicollinearity (high correlation between independent variables).

In [96]:
df['Performance Prediction'] = performance_prediction
df

Timestamp  Kindly Rate your Sleep Quality 😴  \
0   27/10/2023 21:54:15                                 3   
1   28/10/2023 12:24:40                                 4   
2   28/10/2023 12:24:51                                 2   
3   28/10/2023 12:26:11                                 3   
4   28/10/2023 12:26:45                                 2   
5   28/10/2023 12:31:02                                 3   
6   28/10/2023 12:34:45                                 3   
7   28/10/2023 12:35:43                                 4   
8   28/10/2023 12:36:07                                 2   
9   28/10/2023 12:36:20                                 1   
10  28/10/2023 12:37:22                                 2   
11  28/10/2023 12:38:40                                 3   
12  28/10/2023 12:39:43                                 2   
13  28/10/2023 12:40:50                                 4   
14  28/10/2023 12:41:12                                 4   
15  28/10/2023 12:41:56                                 2   
16  28/10/2023 12:48:22                                 3   
17  28/10/2023 13:26:41                                 1   
18  28/10/2023 13:28:47                                 2   
19  28/10/2023 13:29:54                                 2   
20  28/10/2023 13:30:57                                 3   
21  28/10/2023 13:32:41                                 3   
22  28/10/2023 13:33:15                                 2   
23  28/10/2023 13:33:59                                 3   
24  28/10/2023 13:41:28                                 5   
25  28/10/2023 13:42:52                                 3   
26  28/10/2023 13:44:07                                 5   
27  28/10/2023 13:47:44                                 5   
28  28/10/2023 13:52:22                                 2   
29  28/10/2023 13:54:48                                 3   
30  28/10/2023 13:57:20                                 5   
31  28/10/2023 14:09:05                                 5   
32  28/10/2023 14:27:19                                 5   
33  28/10/2023 14:40:55                                 2   
34  28/10/2023 14:58:39                                 4   
35  28/10/2023 15:10:02                                 5   
36  28/10/2023 15:10:43                                 4   
37  28/10/2023 15:39:59                                 3   
38  28/10/2023 16:16:13                                 3   
39  28/10/2023 16:34:16                                 3   
40  28/10/2023 16:55:49                                 4   
41  28/10/2023 17:24:10                                 4   
42  28/10/2023 17:26:18                                 4   
43  28/10/2023 17:29:55                                 2   
44  28/10/2023 17:31:52                                 5   
45  28/10/2023 17:56:58                                 4   
46  28/10/2023 22:33:48                                 2   
47  28/10/2023 23:06:00                                 4   
48  28/10/2023 23:45:09                                 3   
49  29/10/2023 00:32:26                                 1   
50  29/10/2023 11:40:20                                 5   
51  31/10/2023 10:10:03                                 1   
52  31/10/2023 20:49:11                                 3   

    How many times a week do you suffer headaches 🤕?  \
0                                                  1   
1                                                  1   
2                                                  1   
3                                                  2   
4                                                  3   
5                                                  1   
6                                                  5   
7                                                  3   
8                                                  1   
9                                                  2   
10                                                 3   
11                                             

### Please upvote if this was useful 